src/neo4j_graphrag/llm/openai_llm.py

```diff
@@ -155,6 +232,80 @@ async def ainvoke(
         except self.openai.OpenAIError as e:
             raise LLMGenerationError(e)
 
+    async def ainvoke_with_tools(
+        self,
+        input: str,
+        tools: List[dict[str, Any]],  # Tools definition as a list of dictionaries
+        message_history: Optional[Union[List[LLMMessage], MessageHistory]] = None,
+        system_instruction: Optional[str] = None,
+    ) -> ToolCallResponse:
+        """Asynchronously sends a text input to the OpenAI chat completion model with tool definitions
+        and retrieves a tool call response.
+
+        Args:
+            input (str): Text sent to the LLM.
+            tools (List[Dict[str, Any]]): List of tool definitions for the LLM to choose from.
+            message_history (Optional[Union[List[LLMMessage], MessageHistory]]): A collection previous messages,
+                with each message having a specific role assigned.
+            system_instruction (Optional[str]): An option to override the llm system message for this invocation.
+
+        Returns:
+            ToolCallResponse: The response from the LLM containing a tool call.
+
+        Raises:
+            LLMGenerationError: If anything goes wrong.
+        """
+        try:
+            if isinstance(message_history, MessageHistory):
+                message_history = message_history.messages
+
+            params = self.model_params.copy()
+            if "temperature" not in params:
+                params["temperature"] = 0.0
+
+            # Convert tools to OpenAI's expected type
+            openai_tools: List[ChatCompletionToolParam] = []
+            for tool in tools:
+                openai_tools.append(cast(ChatCompletionToolParam, tool))
+
+            response = await self.async_client.chat.completions.create(
+                messages=self.get_messages(input, message_history, system_instruction),
+                model=self.model_name,
+                tools=openai_tools,
+                tool_choice="auto",
+                **params,
+            )
+
+            message = response.choices[0].message
```

Comment:
Not for now because this is also the case in the already existing methods, but we should probably try to reduce code duplication between the sync and async calls.